File: solutions_1-50/problem-24/lexicographic-permutations/src/main.py
import math

def find_permutation(digits, n):
    # Adjust n to be 0-indexed (the 1,000,000th is index 999,999)
    n -= 1
    result = []

    # We use a copy of the digits because we will pop elements out
    available_digits = list(digits)

    for i in range(len(digits) - 1, -1, -1):
        # How many permutations exist for the remaining slots?
        fact = math.factorial(i)

        # Determine which digit belongs in the current position
        index = n // fact
        result.append(available_digits.pop(index))

        # Update n for the next position
        n %= fact

    return "".join(map(str, result))

# Usage for your specific problem:
digits = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
print(find_permutation(digits, 1000000))

File: solutions_1-50/problem-24/lexicographic-permutations/src/main.rs
use factorial::Factorial; // Ensure you ran `cargo add factorial`

fn main() {
    let digits = vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
    println!("{}", find_permutations(digits, 1_000_000));
}

fn find_permutations(digits: Vec<u8>, n: usize) -> String {
    let mut l: u128 = (n - 1) as u128;
    let mut result: Vec<u8> = Vec::new();
    let mut available_digits: Vec<u8> = digits.clone();
    let mut i = (digits.len() as isize) - 1;
    while i >= 0 {
        let fact = (i as usize).factorial() as u128;
        let index = (l / fact) as usize;
        result.push(available_digits.remove(index));
        l %= fact;
        i -= 1;
    }

    return result.iter().map(|&x| x.to_string()).collect();
}
